### DISCUSSION TITLE
Regex versioning not creating update for increased build version

### DISCUSSION THREAD

**Discussion by tophexbit** [renovatebot/renovate#21264] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.27.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitLab Community Edition 15.9.3

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Using these packageRules:
```
"packageRules": [
    {
      "matchPackageNames": ["something-image"],
      "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?-?(?<build>.+)?(-(?<compatibility>.*))$"
    }
  ]
```

Going from version
 `22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python` to
 `22.2.3-22.2.0-20230324142027-3332430-python` should trigger a build update MR.

### What you tried so far.

```
"versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?-?(?<build>.+)?(-(?<compatibility>.*))$"
```

This regex versioning is working, meaning a `patch` update MR is created by Renovate when going from version
 `22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python` to
 `22.2.4-22.2.0-20230324142027-3332430-python` (`patch` version increased from `3` to `4`)
 
 but it's not working, as in Renovate doesn't detect any available updates (I checked the log) when going from version
 `22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python` to
 `22.2.3-22.2.0-20230324142027-3332430-python` (`build` version increased)
 
[I tested the regex (replacing `\\` with `\` before)](https://regex101.com/r/CiMeMJ/1):
it's capturing `22.1.0.p22.1.0-20221111153429-77538bf` and `22.2.0-20230324142027-3332430` as the `build` group as it's supposed to.

So I would expect that `build`
`22.1.0.p22.1.0-20221111153429-77538bf` would be seen as a previous `build` version than
`22.2.0-20230324142027-3332430` and the latter being handled like a new available `build`-update.

Am I missing something? Or is this a bug?

Yes I know it's a horrible versioning scheme (if you can even call it that), but I have to work with that nevertheless 🙂

### Relevant debug logs

<details><summary>Logs</summary>

No Update for build increase:
```
{
  "depName": "something-image",
  "currentValue": "22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python",
  "datasource": "docker",
  "replaceString": "image: something-image:22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python\n",
  "updates": [],
  "packageName": "something-image",
  "warnings": [],
  "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?-?(?<build>.+)?(-(?<compatibility>.*))$",
  "registryUrl": "https://something.com",
  "currentVersion": "22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python",
  "fixedVersion": "22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python"
}
```

Update for patch increase:
```
{
  "depName": "something-image",
  "currentValue": "22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python",
  "datasource": "docker",
  "replaceString": "image: something-image:22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python\n",
  "updates": [
    {
      "bucket": "non-major",
      "newVersion": "22.2.4-22.2.0-20230324142027-3332430-python",
      "newValue": "22.2.4-22.2.0-20230324142027-3332430-python",
      "newMajor": 22,
      "newMinor": 2,
      "updateType": "patch",
      "branchName": "ren-something-image-22-x"
    }
  ],
  "packageName": "something-image",
  "warnings": [],
  "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?-?(?<build>.+)?(-(?<compatibility>.*))$",
  "registryUrl": "https://something.com",
  "currentVersion": "22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python",
  "isSingleVersion": true,
  "fixedVersion": "22.2.3-22.1.0.p22.1.0-20221111153429-77538bf-python"
}
```

</details>

**Comment by secustor** [comment, marked as answer]
There is something wrong with your regex:  https://regex101.com/r/E8Ifsv/1
It detects neither `minor` nor is `build` correct.

**Reply by tophexbit** [reply]
> There is something wrong with your regex: https://regex101.com/r/E8Ifsv/1 It detects neither `minor` nor is `build` correct.

Thanks for the quick answer! You missed replacing one of the `\\` as part of `(\.(?<minor>\\d+))?`. If you replace it with a single `\` it works just like in [my original regex101-link above](https://regex101.com/r/CiMeMJ/1)

So I'm afraid this is not the solution here

**Reply by secustor** [reply]
Yeah you are right, the minor version is now correctly extracted. The build number is still not viable with a value of `22.1.0.p22.1.0-20221111153429-77538bf`. Try a regex which captures `20221111153429` as build number.

**Reply by tophexbit** [reply]
> Yeah you are right, the minor version is now correctly extracted. The build number is still not viable with a value of `22.1.0.p22.1.0-20221111153429-77538bf`. Try a regex which captures `20221111153429` as build number.

I see, I didn't know `build` has to be a number I thought it could also just be a string - but only numbers are supported for `build` yes?

**Reply by viceice** [reply]
yes, build is like  patch, renovate would suggest a patch update.

but renovate will compare the `compabillity` group as string and they need to be equal to allow the new version. if you like to ignore the rest of the version string, just don't capture it. 🙃

**Reply by tophexbit** [reply]
@secustor’s suggestion to just use `20221111153429` as build number should be the best possible solution then

Thank you both 👍
